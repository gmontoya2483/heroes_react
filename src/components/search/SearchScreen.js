import {useForm} from "../../hooks/useForm";
import {getHeroByName} from "../../selectors/getHeroByName";
import {useMemo} from "react";
import HeroCard from "../hero/HeroCard";
import {useNavigate, useLocation} from "react-router-dom";
import queryString from'query-string'

export const SearchScreen = () => {

    const navigate = useNavigate()
    const location = useLocation()


    const { q = '' } = queryString.parse(location.search)



    const [ formValues, handleInputChange, reset ] = useForm({
        searchText: q
    } );

    const { searchText } = formValues
    const filteredHeroes = useMemo(() => getHeroByName(q), [q]);


    const handleSearch = (e) => {
        e.preventDefault();
        navigate(`?q=${ searchText }`)
    }

    return (
        <>
            <h1>Search</h1>
            <hr/>

            <div className="row">
                <div className="col-5">
                    <h4>Search</h4>
                    <hr/>
                    <form onSubmit={ handleSearch }>
                        <input
                            type="text"
                            placeholder={'Buscar un heroe'}
                            className={'form-control'}
                            name="searchText"
                            autoComplete={'off'}
                            value={ searchText }
                            onChange={ handleInputChange }
                        />

                        <div className="d-grid gap-2  mt-1">
                            <button
                                type={'submit'}
                                className={'btn btn-outline-primary'}
                            >
                                Buscar...

                            </button>
                        </div>

                    </form>
                    
                </div>
                <div className="col-7">
                    <h4>Resultados</h4>
                    <hr/>

                    {
                        (q === '')
                            ? <div className={'alert alert-info'}>Buscar un Heroe... </div>
                            : (filteredHeroes.length === 0) && <div className={'alert alert-danger'}>No hay resultados: { q }</div>
                    }

                    {
                        filteredHeroes.map( (hero) =>(
                            <HeroCard
                                key={hero.id}
                                { ...hero }
                            />
                        ))
                    }



                </div>
            </div>

        </>
    );
};
