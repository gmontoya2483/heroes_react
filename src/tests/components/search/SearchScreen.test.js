import {SearchScreen} from "../../../components/search/SearchScreen";
import {mount} from "enzyme";
import {MemoryRouter} from "react-router-dom";


const mockNavigate = jest.fn();

jest.mock('react-router-dom', ()=> ({
  ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}));


describe('Pruebas en <Search Screen />', ()=> {

    test('Se debe de mostrar correctamente con valores por defecto', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar un Heroe...')


    });


    test('debe de mostrar a Batman y el input con el valor del queryString', ()=> {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect( wrapper.find('input').prop('value').trim()).toBe('batman');

    });


    test('debe de mostrar un error si no existe el superheroe', ()=> {

        const searchTerm = 'yuyuyuy'

        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${ searchTerm }`]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-danger').text().trim()).toBe(`No hay resultados: ${ searchTerm }`)

    });


    test('se debe de llamar a funcion navigate al nuevo url', ()=> {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault: () => {}
        });

        expect( mockNavigate ).toHaveBeenCalledWith('?q=batman');


    });


});
