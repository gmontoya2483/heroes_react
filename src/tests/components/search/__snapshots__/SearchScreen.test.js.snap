// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Search Screen /> Se debe de mostrar correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un heroe"
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2  mt-1"
        >
          <button
            className="btn btn-outline-primary"
            type="submit"
          >
            Buscar...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un Heroe... 
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <Search Screen /> debe de mostrar a Batman y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un heroe"
          type="text"
          value="batman"
        />
        <div
          className="d-grid gap-2  mt-1"
        >
          <button
            className="btn btn-outline-primary"
            type="submit"
          >
            Buscar...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="img-fluid rounded-start"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                   
                  Batman
                   
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Mas ...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <Search Screen /> debe de mostrar un error si no existe el superheroe 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un heroe"
          type="text"
          value="yuyuyuy"
        />
        <div
          className="d-grid gap-2  mt-1"
        >
          <button
            className="btn btn-outline-primary"
            type="submit"
          >
            Buscar...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No hay resultados: 
        yuyuyuy
      </div>
    </div>
  </div>,
]
`;
